2024/day4/day4.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	_ "embed"
+	"fmt"
+	"strings"
+)
+
+//go:embed test.txt
+var test string
+
+//go:embed input.txt
+var input string
+
+func main() {
+	board := parseInput(input)
+	xmasCount := wordSearch(board, "XMAS")
+	fmt.Printf("%d\n", xmasCount)
+	o := xMAS(board)
+	fmt.Printf("xMAS Count: %d\n", o)
+}
+
+func xMAS(board [][]string) int {
+	checkMAS := func(i, j int) int {
+		if i >= len(board) || j >= len(board[0]) || i < 0 || j < 0 {
+			return 0
+		}
+
+		switch board[i][j] {
+		case "S":
+			return 1
+		case "M":
+			return 2
+		}
+		return 0
+	}
+
+	count := 0
+	for i := range len(board) {
+		for j := range len(board[0]) {
+			if board[i][j] == "A" {
+				if (checkMAS(i+1, j+1)+checkMAS(i-1, j-1) == 3) && (checkMAS(i+1, j-1)+checkMAS(i-1, j+1) == 3) {
+					fmt.Println(i, j)
+					count++
+				}
+			}
+		}
+	}
+
+	return count
+}
+
+func wordSearch(board [][]string, word string) int {
+	count := 0
+
+	var search func(x, y, idx int, path [][2]int, xDir int, yDir int)
```

Comment:
Interestingly. cannot recursivly call in a nested func, unless u explicitly declare it before. Which make sense i guess.